### ISSUE TITLE
TLS cert per site

### ISSUE THREAD

**Issue by stokito** [cunnie/sslip.io#22]
#2 was a bad situation and very questionable because nobody was harmed. OK, potentially some user might think that his interaction with site remains confidential.
It's still better than plain http...

I found the http://local-ip.co and they giving the cert and looks like it wasn't revoked. Or even if revoked looks like noone stops from issuing a new one. 

But I think that if the new cert will be generated for each site then it will be totally legal.

It's just the same as a company's trusted subcontractor webstudio generates the cert for you.

It is possible to have many signed certs for the same domain.
I checked and Letsencrypt have huge rate limits.
A website can call with a simple wget a special url like https://sslip.com/mycert.pem and the new cert/key is signed and returned specifically for the remote ip.
E.g to get a cert for 1.1.1.1.sslip.com you have to call it from that IP. This gives a good protection.

Since you downloaded the cert over https then only the sspip knows the private key. It's not compromised and publicly known.

Of course if the IP is belongs to ISP and many people use it behind CGNAT then anybody else can generate a cert for the same domain and make MiTM but they can't just passivelly listen and decrypt. The attack vector is significantly reduced.

In fact here the site may send a Certificate Sign Request e.g. the sspip will act as a CA. Then the sspio can't decrypt recorded trafik. But this is overcomplication.

But even the MiTM can be prevented by introducing registration. Nobody can't claim that they own an IP adress forewer. Then use a four level domain as a username. E.g.
alice.1.1.1.1.sslip.com 
If Alice took the sub.sub.domain then if anyone want to issue a cert for it they have to know her psssword.

Right now for my server I can generate a cert for it's static ip with a Letsencrypt certbot.
The problem is that for my router behind CGNAT it's not possible because the port 80 is closed.
With upnp I can forward some ephemeral port for the webserver of my router. So I can't run the certbot with http verification and I need for the DNS verification which is something that only the sspip can make.
If if some router is not behind a NAT then still the certbot or even acme.sh are too heavy for a router.
So it would be really nice if the sslip can generate and sign a cert.

**Comment by cunnie** [comment]
>  I found the http://local-ip.co/ and they giving the cert and looks like it wasn't revoked. Or even if revoked looks like no one stops from issuing a new one.

I've had two certs revoked for doing exactly what local-ip.co is doing (publishing the private key), and I've learned my lesson: I won't do it a third time. I promised the CTO of Sectigo that I *wouldn't publish the key* if he relented and issued me a replacement cert. He kept his part of the bargain, and I intend to keep mine.

I can see you're in a tough position; the best I can offer is [this procedure](https://github.com/cunnie/sslip.io/blob/main/docs/wildcard.md). I wish I had a better solution.
